src/nvim/optionstr.c

```diff
@@ -324,30 +324,32 @@ void set_string_option_direct(const char *name, int opt_idx, const char *val, in
     }
   }
 
-  if (is_hidden_option(idx)) {       // can't set hidden option
+  vimoption_T *opt = get_option(idx);
+
+  if (opt->var == NULL) {  // can't set hidden option
     return;
   }
 
-  assert((void *)get_option_var(idx) != (void *)&p_shada);
+  assert((void *)opt->var != (void *)&p_shada);
 
   s = xstrdup(val);
   {
-    varp = (char **)get_option_varp_scope(idx, both ? OPT_LOCAL : opt_flags);
-    if ((opt_flags & OPT_FREE) && (get_option_flags(idx) & P_ALLOCED)) {
+    varp = (char **)get_varp_scope(opt, both ? OPT_LOCAL : opt_flags);
+    if ((opt_flags & OPT_FREE) && (opt->flags & P_ALLOCED)) {
       free_string_option(*varp);
     }
     *varp = s;
 
     // For buffer/window local option may also set the global value.
     if (both) {
-      set_string_option_global(idx, varp);
+      set_string_option_global(opt, varp);
     }
 
-    set_option_flag(idx, P_ALLOCED);
+    opt->flags |= P_ALLOCED;
 
     // When setting both values of a global option with a local value,
     // make the local value empty, so that the global value is used.
-    if (is_global_local_option(idx) && both) {
+    if ((opt->indir & PV_BOTH) && both) {
```

Comment:
We lost some readability here. I assume that you removed `is_global_local_option` and related functions because they just aren't that common, so the code reduction might be worth it.

Reply:
Fyi, this is also what `option.c` does. `option.c` directly checks `indir & PV_BOTH` instead of using `is_global_local_option`, that was my primary reasoning for removing it